### DISCUSSION TITLE
How do I design my app with database access so it can run in desktop and web?

### DISCUSSION THREAD

**Discussion by heejit** [cogentcore/core#1237] [General]
when I am launching my app with `core run web` it shows following on web browser and error in web console I understand that
wasm app have some restriction so how do I design my app.

![image](https://github.com/user-attachments/assets/b76fd1aa-1def-4e96-8054-b787f4f1b7bc)

**Comment by kkoreilly** [comment]
You have two options: use something like https://github.com/goki/rqlite, which works on the web, or send requests to a separate server that handles database requests with an API, which is what you would do with a normal web app.

**Reply by heejit** [reply]
It required web server with https. 

I try with following to generate a certificate 
go run ~/sdk/go1.23.0/src/crypto/tls/generate_cert.go --host="localhost"

Run web server with generate certificate

App shows :
ErrSendingRequest(3):  error sending request, err: Get "https://127.0.0.1:8080/myip": net/http: fetch() failed: TypeError: NetworkError when attempting to fetch resource.(Please check your network, maybe use bad proxy or network offline)

Webserver shows following error:
http: TLS handshake error from 127.0.0.1:55524: remote error: tls: bad certificate

But same application running fine on desktop.

**Reply by kkoreilly** [reply]
What package are you using in the app to do this? Please tell me what package you are using and then I can help you figure out what is going wrong. Ideally you could paste in a snippet of the code that is causing the error. Thank you!

**Reply by heejit** [reply]
Here is the full code:

generate certificate:
go run ~/sdk/go1.23.0/src/crypto/tls/generate_cert.go --host="localhost"

**server/main.go**
```
package main


import "os"
import "net/http"
import "log"
import "github.com/gin-gonic/gin"


func main() {
	var GinRouter *gin.Engine

	gin.SetMode(gin.ReleaseMode)
	GinRouter = gin.Default()
	GinRouter.GET("/myip", IpHandler)
	err := GinRouter.RunTLS(":8080", "cert.pem", "key.pem")
	if err != nil {
		log.Fatalln(err)
	}
}


func IpHandler(c *gin.Context) {
	c.String(http.StatusOK, c.ClientIP()+"\n")
}

```

**client/main.go**
```
package main

import (
	"cogentcore.org/core/core"
	"cogentcore.org/core/events"
	"github.com/go-zoox/fetch"
)

func main() {
	b := core.NewBody("Test")
	t := core.NewTextField(b)
	btn := core.NewButton(b)
	btn.SetText("Call")
	btn.OnClick(func(e events.Event) {
		var vErr error
		var vResp *fetch.Response
		vResp, vErr = fetch.Get("https://127.0.0.1:8080/myip", &fetch.Config{TLSInsecureSkipVerify:true})
		if vErr != nil {
			t.SetText(vErr.Error())
		}

		if vErr == nil {
			t.SetText(vResp.String())

		}
		t.Update()

	})
	b.RunMainWindow()
}
```

core run web -port=9000
Serving at http://localhost:9000

curl https://127.0.0.1:8080/myip --insecure

**Reply by kkoreilly** [reply]
Typically you do not use `https` for localhost; try just using `http://127.0.0.1` for everything and no `tls` stuff and then it should work. You shouldn't have to manage `tls` yourself, and the browser won't like that, which is why it's failing.

**Reply by heejit** [reply]
Following errors  after change to http no tls:

client app in browser:
`ErrSendingRequest(3):  error sending request, err: Get "http://127.0.0.1:8080/myip": net/http: fetch() failed: TypeError: NetworkError when attempting to fetch resource.(Please check your network, maybe use bad proxy or network offline)`

Server app if client running in browser:
`[GIN] 2024/09/26 - 20:57:17 | 404 |       1.859µs |       127.0.0.1 | OPTIONS  "/myip"`

Server app if client running in desktop:
`[GIN] 2024/09/26 - 20:56:26 | 200 |      26.425µs |       127.0.0.1 | GET      "/myip"`


Why server shows OPTION instead GET

**Reply by kkoreilly** [reply]
`OPTIONS` is a pre-request that gets capabilities. All web browsers send it along with requests, so you need to support it on your server. https://github.com/gin-gonic/gin/issues/559 might be informative.

**Reply by heejit** [reply]
I try gin contrib cors not work try few other from internet not work only following works:


```
https://asanchez.dev/blog/cors-golang-options/

func CORS(c *gin.Context) {
	// First, we add the headers with need to enable CORS
	// Make sure to adjust these headers to your needs
	c.Header("Access-Control-Allow-Origin", "*")
	c.Header("Access-Control-Allow-Methods", "*")
	c.Header("Access-Control-Allow-Headers", "*")
	c.Header("Content-Type", "application/json")

	// Second, we handle the OPTIONS problem
	if c.Request.Method != "OPTIONS" {
		c.Next()
	} else {
		// Everytime we receive an OPTIONS request,
		// we just return an HTTP 200 Status Code
		// Like this, Angular can now do the real
		// request using any other method than OPTIONS
		c.AbortWithStatus(http.StatusOK)
	}
}
```

before every api call browser is calling OPTIONS

My operating system is Opensuse 
firefox -> calling options
chormium -> is not calling options

**Reply by heejit** [reply]
Thanks for your time kkoreilly
